Clean up text for Client-DP (#22)
wangshan/draft-wang-ppm-differential-privacy: draft-wang-ppm-differential-privacy.md

> TODO: Specify a Gaussian sampler

## Binary Randomized Response {#rr}

The binary randomized response mechanism specified here aims to provide the

> DP guarantee at the Client level. It is the same as the one specified in

Reviewer: Be specific ("approximate-DP" is more meaningful than "DP guarantee" because we defined it previously) and specify the trust model.
```suggestion
good approximate-DP in the OAMC trust model. It is the same as the one specified in
```

Author: It actually is `EPSILON_0`-DP, but I removed "DP guarantee", and say it is `EPSILON_0`-DP in the deletion-DP model.

Appendix C.1 of {{MJTB+22}}.

To apply noise at Client side, the Client flips a coin at each coordinate of its

> vector. If the coin turns out to be a head, the Client will change the value `x`

Reviewer: We need a paragraph before this describing what type of data the mechanism is compatible with. Giving an illustrative example would help to.

Author: I added a paragraph, and reworked the explanation here. Please take a look again.

at that coordinate to `1 - x`. The probability of the coin giving a head is
`1 / (e^(EPSILON_0) + 1)`, where `EPSILON_0` provides the pure `EPSILON_0`-DP

> guarantee, in the deletion-DP model (Definition II.4 of {{EFMR+20}}), which

Reviewer: 1. If you say the words "DP guarantee", then you probably aren't being clear enough. Do you mean that we get `EPSILON_0`-DP in the OC trust model?
2. In the appendix we say that DAP is not compatible with deletion-DP. Indeed, the way we define neighboring databases in the section above rules this out.

Reviewer 2: for 1. I think it's ok to see EPSILON_0-DP guarantee, since this is client level DP noise, it doesn't really to do with the DP guarantee for the whole aggregation, which then relates to the trust models.

Author: So it's correct binary randomized response only provides `EPSILON_0`-DP in deletion-DP at Client level, but by aggregating a batch of one-hot vectors noised with binary randomized response at all coordinates, we get `(EPSILON, DELTA)`-DP in the replacement DP model. I reworked this paragraph a bit. Let me know if it makes sense more.

ensures that it is impossible to confidently determine whether an individual

> contributed data to the final aggregate result.

Reviewer: "which ensures that it is impossible..." What's the goal for this text?

Author: Removed it.

The final aggregate result is obtained by "debiasing" the aggregation of the
noisy Client measurements. It is expressed as a vector of floats, because the

> debiasing operations involve floating point arithmetics.

Reviewer: ```suggestion
debiasing operations involve floating point arithmetic.
```

Since the noise generated by each Client at each coordinate is independent, and
based on a coin flip with probability `1 / (e^(EPSILON_0) + 1)`, the noise
generated by `n` number of Clients at each coordinate is a binomial

> distribution. By Central Limit Theorem (CLT), as `n` grows, the noise

Reviewer: CLT is probably folklore enough ("folklore" == "well-known enough to not require a citation") but we should probably state it for the reader in the conventions section of this draft, at least informally. (Similar to how we handle `e`.)

Author: Sounds good. I added a text about CLT in Conventions.

> TODO: Specify a Gaussian sampler

## Binary Randomized Response {#rr}

> 

Reviewer: This isn't suitable as an intro paragraph. I suggest moving it below the paragraph that mentions approximate DP and OAMC trust model.

Author: I opened up this section with binary randomized response's parameters and inputs right now. I mention `EPSILON_0`-DP in OC trust model now, which I think we've already defined at this point?

Reviewer: Looks much better, thanks!

> The binary randomized response mechanism specified in Appendix C.1 of
> {{MJTB+22}} is `EPSILON_0`-DP at the Client level, in the deletion-DP model
> (Definition II.4 of {{EFMR+20}}).

Reviewer: (Follow up on https://github.com/wangshan/draft-wang-ppm-differential-privacy/pull/22#discussion_r1364371532) This looks fine except "Client level" is not defined. What about "in the OC trust model"?

Also, we should define deletion-DP informally here so that the reader doesn't need to look this up.

```suggestion
The binary randomized response mechanism specified in Appendix C.1 of
{{MJTB+22}} is `EPSILON_0`-DP at the OC trust model, in the deletion-DP model
(Definition II.4 of {{EFMR+20}}).
```

Author: Yes, good idea. It indeed works under OC trust model.

> Binary randomized response's input is a single bit `x`. The bit is flipped to

Reviewer: A single bit or a bit vector? I still think an illustrative example would be helpful. (What type of data might be noising? Perhaps describe the histogram use case?)

Author: The vanilla binary randomized response operates on a bit. I described it with an example with a bit, let me know if it makes things clearer.

I did mention in the next paragraph: "we generalize binary randomized response mechanism by applying it independently at all coordinates of a Client's bit vector".

Reviewer: Yup, that clarifies it, thanks.

`1 - x` with probability `1 / (exp(EPSILON_0) + 1)`.

We generalize this noise by applying it independently at all coordinates of a
Client's bit vector. Under OAMC trust model, it is proven in Appendix C.1.3

> of {{MJTB+22}} that we get good `(EPSILON, DELTA)`-DP in the replacement-DP

Reviewer: "replacement-DP" is defined below (in the appendix) but not above.  If you want to use the term, make sure it's defined before reader gets here. (I don't think it's strictly necessary, since we defined DP in terms of replacement already.)

Author: Good point. Move that part of the text to overview now, where we've been defining `EPSILON`-DP and `(EPSILON, DELTA)`-DP in terms of replacement-DP. I didn't go in depth on deletion-DP, which I only elaborated in the section for binary randomized response.

> TODO: Specify a Gaussian sampler

## Binary Randomized Response {#rr}

> 

Reviewer: nit: A better intro sentence would be: "This section described the binary randomized response mechanism."

parameter to its measurement, the Client gets `EPSILON_0`-DP in deletion-DP
model (Definition II.4 of {{EFMR+20}}, and Definition C.1 of {{MJTB+22}}).
A formal definition of deletion-DP is elaborated in {{rr-deletion-dp}}.

We generalize binary randomized response mechanism by applying it independently
at all coordinates of a Client's bit vector. Under OAMC trust model, it is
proven in Appendix C.1.3 of {{MJTB+22}} that we get good `(EPSILON, DELTA)`-DP
in the replacement-DP model, by aggregating a batch of noisy Client
measurements, each of which is a bit vector with exactly one bit set, and is
noised with binary randomized response at all coordinates. The final aggregate
result needs to be "debiased" due to the noise added by the Clients, and is
expressed as a vector of floats, because of floating point arithmetic.

Since the noise generated by each Client at each coordinate is independent, and
as the number of Clients `n` grows, the noise distribution at each coordinate
approximates a Gaussian distribution, with standard deviation
`sqrt(n * exp(EPSILON_0) / (exp(EPSILON_0) - 1)^2)`, as proved by Theorem C.2 of
{{MJTB+22}}.

> ### `EPSILON_0`-DP in Deletion-DP {#rr-deletion-dp}

Reviewer: This section is well-scoped, and I don't want to change anything for now.

In the medium term, I think we should spend time thinking through how this notion of DP relates to the notion of DP we've already defined.

Long term, I think our goal for PPM should be to have a single (or as few as possible) notions of DP. That way we have a standard methodology (or methodologies) for evaluating DP mechanisms and policies. The fact that there are so many DP notions in the literature that may or may not be comparable is a huge problem for standardization.

Relatedly, I believe there is an important research question here: for a given application (DAP) systematize the DP notions in the literature that apply such that we can compare them to one another (at least qualitatively, but ideally quantitatively). Presumably there are trade-offs to be made, depending on which notion of DP you pick; what we want at the end of the day is the right DP notion that allows us to make sense of what to do for DAP.

Author: Agreed. Let's continue discussing in issue.

> JC: We only add a definition of deletion-DP here since this is likely the only
> mechanism that provides deletion-DP in the OC trust model. Putting it in
> overview might confuse readers early on, because we said only replacement-DP
> applies to DAP.

Let `P1(S, D, E)` denote the probability that a randomized algorithm `S`, on an
input measurement `D`, outputs a noisy measurement `E`. Let `P2(R, E)` denote
the probability that a reference noisy output `R` is equal to `E`. A  randomized
algorithm `S` is said to be `EPSILON_0`-DP in the deletion-DP model, if there
exists a reference distribution `R`, such that for all possible Client
measurements `D`, and all possible noisy outputs `E`, we have:

~~~
-EPSILON_0 <= ln(P1(S, D, E) / P2(R, E)) <= EPSILON_0
~~~

Intuitively, if we think of the reference distribution `R` as an average, noisy
Client measurement, deletion-DP makes it hard to distinguish the Client

> measurement `D`, from the measurement from an average Client.

Reviewer: ```suggestion
measurement `D` from the measurement from an average Client.
```

  `DpMechanism.debias`).

The composition of Client- and Aggregator-DP mechanisms defines the DP

> policy for a VDAF, and enforces the target central DP guarantee.

Reviewer: "central DP guarantee" is not meaningful unless we define it.

Author: Removed this change.

## Histograms

Many applications require aggregating histograms in which each Client submits a
bit vector with exactly one bit set, also known as, "one-hot vector". We
describe two policies that achieve central `(EPSILON, DELTA)`-DP on this use
case: one which uses only Client-DP mechanism and targets the OAMC trust model,

> and another which uses only Aggregator-DP mechanism and targets the more

Reviewer: ```suggestion
and another which uses only an Aggregator-DP mechanism and targets the more
```

> stringent OAOC trust model. We will then compare the utilities of both policies

Reviewer: ```suggestion
stringent OAOC trust model. We will then compare the utility of both policies
```

in different settings of `EPSILON` and `DELTA`, which can be tuned to provide
different levels of DP guarantee.

### Prio3MultiHotHistogram with Client-DP

> > TODO Figure out if there is a way to endow Prio3Histogram with local DP

Reviewer: This TODO looks to be out-of-date.
